cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -291,6 +351,79 @@ public void validateProperties(String provisionerName, Map<String, String> prope
     provisioner.validateProperties(properties);
   }
 
+  private Runnable createProvisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner) {
+    ProgramRunId programRunId = taskInfo.getProgramRunId();
+    ProvisionerContext context =
+      new DefaultProvisionerContext(programRunId, taskInfo.getProvisionerProperties(),
+                                    sparkCompat, createSSHContext(taskInfo.getSecureKeyInfo()));
+
+    // TODO: (CDAP-13246) pick up timeout from profile instead of hardcoding
+    ProvisioningTask task = new ProvisionTask(taskInfo, transactional, datasetFramework, provisioner, context,
+                                              provisionerNotifier, 300);
+    return () -> {
+      try {
+        task.execute();
+      } catch (InterruptedException e) {
+        LOG.debug("Provision task for program run {} interrupted.", taskInfo.getProgramRunId());
+      } catch (Exception e) {
+        LOG.info("Provision task for program run {} failed.", taskInfo.getProgramRunId(), e);
+      }
+    };
+  }
+
+  private Runnable createDeprovisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner,
+                                         Consumer<ProgramRunId> taskCleanup) {
+    Map<String, String> properties = taskInfo.getProvisionerProperties();
+    ProvisionerContext context =
+      new DefaultProvisionerContext(taskInfo.getProgramRunId(), properties,
+                                    sparkCompat, createSSHContext(taskInfo.getSecureKeyInfo()));
+    DeprovisionTask task = new DeprovisionTask(taskInfo, transactional, datasetFramework, 300,
+                                               provisioner, context, provisionerNotifier, locationFactory);
+    return () -> {
+      try {
+        task.execute();
+        taskCleanup.accept(taskInfo.getProgramRunId());
+      } catch (InterruptedException e) {
+        // We can get interrupted if the task is cancelled or CDAP is stopped. In either case, just return.
+        // If it was cancelled, state cleanup is left to the caller. If it was CDAP master stopping, the task
+        // will be resumed on master startup
+        LOG.debug("Deprovision task for program run {} interrupted.", taskInfo.getProgramRunId());
+      } catch (Exception e) {
+        // Otherwise, if there was an error deprovisioning, run the cleanup
+        LOG.info("Deprovision task for program run {} failed.", taskInfo.getProgramRunId(), e);
+        taskCleanup.accept(taskInfo.getProgramRunId());
+      }
+    };
+  }
+
+  private List<ProvisioningTaskInfo> getInProgressTasks() {
+    return Retries.callWithRetries(
+      () -> Transactionals.execute(transactional, dsContext -> {
+        ProvisionerDataset provisionerDataset = ProvisionerDataset.get(dsContext, datasetFramework);
+        return provisionerDataset.listTaskInfo();
+      }),
+      co.cask.cdap.common.service.RetryStrategies.exponentialDelay(10, 120, TimeUnit.SECONDS),
```

Comment:
It seems like you are not interrupting the start up thread when someone call stop on this service. This means the stopping could be blocked for as long as 120 seconds?

Reply:
i don't see a way to get the startup thread in AbstractIdleExecutor. I'll reduce the sleep, but log less often.

Reply:
There is a way, but not so easy. I think just reducing the sleep is fine. To actually make it more reactive to stop, you have to do something like this:

```java
  public static final class SimpleService extends AbstractIdleService {

    private volatile Thread startThread;

    public SimpleService() {
      addListener(new ServiceListenerAdapter() {
        @Override
        public void stopping(State from) {
          if (from == State.STARTING && startThread != null) {
            startThread.interrupt();
          }
        }
      }, Threads.SAME_THREAD_EXECUTOR);
    }

    @Override
    protected void startUp() throws Exception {
      startThread = Thread.currentThread();
      if (state() != State.STARTING) {
        // Stop has issued.
        return;
      }
      TimeUnit.SECONDS.sleep(1000);
    }

    @Override
    protected void shutDown() throws Exception {
      System.out.println("Shutdown");
    }
  }
```